### DISCUSSION TITLE
Hayward Turbocell

### DISCUSSION THREAD

**Discussion by celestinjr** [tagyoureit/nodejs-poolController#921] [Q&A]
Will this project support Hayward salt cells?  I see reference to “Intellichlor, Aqua-Rite and OEM brands.”

My particular model is a Hayward Turbocell TCELL940.

**Comment by tagyoureit** [comment]
Yes, we support it.  [Aqua Rite](https://hayward.com/aquarite-900-w-40-000-gallon-extended-life-turbocell-aqr940.html) is a Hayward product.    [TurboCell](https://hayward.com/turbocell-extended-life-salt-cell-40-000-gallons-tcell940.html) is the name for the chlorine generation cell specifically.

**Reply by celestinjr** [reply]
Got it.  Does this project support direct communication to the cell, or does it rely on RS-485 communication via an OEM salt cell controller?

I have an Aqualogic AQL-PS-8 currently, but I was imagining this project could replace that entirely. 

Appreciate everyone’s efforts to create what looks like a fantastic project!

**Reply by tagyoureit** [reply]
You can definitely replace the entire controller.  Our project works based on RS-485.  There have been a number of people that have converted their exist OCP (outdoor control panel) entirely over to Nixie.  It's useful to keep the guts (relays, power supplies, etc) as you need those items anyway.  

It's usually helpful to keep the salt cell controller because it provides the right power requirements to the cell, but it doesn't do anything special with the rs-485 messages other than be a middleman.

**Reply by celestinjr** [reply]
Reading through the wiki a few times, I think I’ve rationalized how to connect my equipment (valves, pumps, sensors, etc.) to Nixie with the exception of my salt cell.  It appears I can rewire the other sensors and relays to the appropriate Pi HATs.  I would be able to reuse one of my existing AQL-PS-8 relays for the booster pump, but the relay signal would be rewired to the Pi relay hat instead of the original AQL board.

When you refer to keeping “the salt cell controller”, are you referring to the entire AQL-PS-8?  Would I retain power to the AQL board, and the Nixie would be able to control the salt cell via RS-485 through the AQL-PS-8?  How would I avoid a conflict with the AQL board wanting to control the cell?  For instance, if the AQL isn’t detecting flow (because I moved the flow switch to the Nixie), I expect the AQL will be sending a 0% signal to the salt cell. 

I feel I’m probably missing something here.  I’m unclear if there’s a means of segregating RS-485 communication to the salt cell from the rest of the AQL board logic.

**Reply by tagyoureit** [reply]
First, good observation on the booster relay.  Never have an undersized relay control directly control power to the pumps.  

For the power requirements for the cell, you might want to look through this [troubleshooting manual](https://ca.hayward.com/media/akeneo_connector/asset_files/A/Q/AQR159d_AquaRite_TSG_8442.pdf).  Starting around page 32 it shows the power that would be coming from the standalone chlorinator controller (~24vdc).

**Reply by celestinjr** [reply]
In my case I have an AquaLogic panel ([board shown here](https://www.troublefreepool.com/wiki/index.php?title=Hayward_Aqualogic#Hayward_AquaLogic)), which appears to be a lot of the same components in the AquaRite you linked to, but my panel has other automation features.  I actually found that I had the equivalent troubleshooting guide that you posted for my panel already stashed away, so I'm now clear on where the salt cell is getting power from.

I've gathered that the Nixie requires some type of appropriately powered intermediary (logic board) to speak to and power the salt cell.  In my case, I'm assuming that would have to be my existing AQL-PS-8 AquaLogic board communicating to the Nixie via RS-485.  What I'm unclear on is what functionality (if any) the Nixie will have for controlling the salt cell (on/off, output percentage, super chlorinate, etc.)

Questions (assuming my AquaLogic board w/ connected salt cell has power and is connected to the Nixie via RS-485):
1. Can the Nixie turn the salt cell on/off at will?
2. If yes to (1), does the Aqualogic need to retain flow detection or is that all handled by Nixie?
3. Can the Nixie control cell output percentage (without navigating through the AquaLogic menus to change the setting)?
4. Can the Nixie access the water temperature and/or salt content reported by the salt cell?

Currently, I have a schedule set in the AquaLogic panel to run the pump in pool/spa/spillover, during which the salt cell runs at the current pool or spa chlorinator output level setting.  I'm unclear how or whether the Nixie would overrule the schedule in the AquaLogic, which is one of (but not the only) driver to switch over.

**Reply by tagyoureit** [reply]
You can power the SWG with 18-33vdc and likely need to provide 10-12v on the RS485 bus (per other threads [one](https://github.com/tagyoureit/nodejs-poolController/discussions/439), [two](https://github.com/tagyoureit/nodejs-poolController/issues/293) and [three](https://github.com/tagyoureit/nodejs-poolController/discussions/415).

1. Can the Nixie turn the salt cell on/off at will?
Yes. 

2. If yes to (1), does the Aqualogic need to retain flow detection or is that all handled by Nixie?
The chlorinator, be design, will not be generating chlorine when the main pump is off but sometimes the flow sensor goes bad.  When you wire it up, make sure that it is _only_ powered when your primary pool pump is running.  This is standard best practice.  Aqualogic doesn't know about flow detection (it's purely in the cell but can report 'no flow' status).  It only tells the cell to turn on or off at the specified output %.

3.  Can the Nixie control cell output percentage (without navigating through the AquaLogic menus to change the setting)?
Yes.  Forget about using the Aqualogic and get a standalone power supply.  I thought you might be able to use it for power but it's not worth it to keep it in the mix.

4. Can the Nixie access the water temperature and/or salt content reported by the salt cell?
Salt level, yes (although these cells are pretty inaccurate).  Not water temperature.  Some cells will report 'low water temp' but none of the cells report their water temp back to the controller.

When Nixie has control of the salt cell you can set the pool (or spa) set point and salt target.

![2024-04-02_17-48-28](https://github.com/tagyoureit/nodejs-poolController/assets/1646840/86d77421-de5e-43ca-a77f-c8b5e6f73198)

If you decide to do add an ORP sensor in the future, REM Chem can control the chlorine generation of the cell based on an ORP target.

**Reply by celestinjr** [reply]
Thanks for suffering through this with me.  I'll catch on, I promise!  I think some of this may be terminology I'm struggling with.

In my setup, what are you referring to as the SWG?  I'm assuming the SWG is a portion of the AquaLogic control board responsible for controlling the salt cell, since I don't have a separate SWG board.  If you look closely at the [AquaLogic board](https://www.troublefreepool.com/wiki/index.php?title=File:Hayward_GLX-PCB-MAIN_Board.jpg), it has all of the components of the AquaRite (power supply, rectifiers, etc.) and then some.

The salt cell pinout ([shown here](https://www.troublefreepool.com/wiki/images/thumb/8/84/Hayward_T-Cell_Cable_Pinout.jpg/600px-Hayward_T-Cell_Cable_Pinout.jpg)) implies that the cell itself has no RS-485 capabilities, so I'm in need of more than just a power supply for communication with the Nixie, right?  When you say forget about the AquaLogic and get a standalone power supply, are you referring to a standalone power supply that also translates RS-485 to power and obtains salt level from the cell?

**Reply by tagyoureit** [reply]
SWG = salt water generator.  Just an acronym for the salt system.

I was going off my knowledge of the Intellichlor cells.  So yes, you'll need AQL-PS-8.  You can either start by hooking up njsPC to that box and see how far you can get (you can extend the number of relays, etc).  A much more involved route could be to set up 2 parallel systems; one with the AQL-PS-8 and another Nixie.  In this case you could have Nixie turn on the AQL-PS-8 pool circuit/chlorinator and there would need to be a bridge (here's [one example](https://github.com/tagyoureit/nodejs-poolController/discussions/413) or [another](https://github.com/tagyoureit/nodejs-poolController/discussions/527).  The feeds/triggers could be a way to do that.

**Reply by celestinjr** [reply]
Today I got the MEGABAS connected to the RS485 port on my AQL-PS-8, but I'm not having any luck thus far.  I know that the AQL-PS-8 panel communicates at 19.2k baud rate, but I don't receive any successful messages unless the baud rate is set to 9600.  When using a baud rate of 9600, I see only one repeated inbound message:

![image](https://github.com/tagyoureit/nodejs-poolController/assets/30021355/c07ea295-4ab9-410f-a092-a614502bffaa)

I haven't made sense of what this is yet.

Note that I have a separate custom built device I've been using to control my AQL-PS-8 panel via RS485, so I know it's communicating at 19.2k, and I haven't been able to make a connection to this repeated message that MessageManager is reporting.

**Reply by tagyoureit** [reply]
You can't have different baud rates on the same bus.  I'd try removing your custom built device so there is not interference on the lines.

**Reply by celestinjr** [reply]
Agreed, which is why I initially tried 19.2k. I’ve confirmed no change with my other device connected or not.

When set at 19.2k and 2 stop bits, I see numbers increasing under “received” but “successful” remains zero. 

I found [this](https://github.com/tagyoureit/nodejs-poolController/issues/393) discussion which appears to have successfully communicated with a Hayward pump using what appears to be the same 19.2k baud rate / 2 stop bit setup.  Not sure why I’m not seeing any messages.

**Reply by tagyoureit** [reply]
Are you trying to connect all the equipment at the same time?  ie the chlorinator and the PS-8 panel and/or the pump?  If so, try to communicate with just one of them at a time.  Seems they may each have different settings.

**Reply by celestinjr** [reply]
Nope. The only device on the bus is the AquaLogic automation panel (AQL-PS-8). 

Surprisingly, I left the MessageManager running overnight, and there was a short period of time that it captured some messages. They don’t appear to have been captured correctly, but it’s a start.  Based on knowing what the messages should look like, I see legitimate data, but it’s not being split correctly. I see pieces of messages in the “padding”. 

I need to spend more time digging through what was captured and figuring out what local time they correspond to. I can post more once I do that.

**Reply by tagyoureit** [reply]
Double check the wiring.  19.2k, no parity, 2 stop bits should result in the same messages whether it's you have your other device connected to the control panel or njsPC.

**Reply by celestinjr** [reply]
Alright so after additional troubleshooting and reversing A/B RS485 wiring countless times with no success, I ran 'cat /dev/ttyS0' and to my surprise saw all of the data (in ASCII format of course).  This seems to confirm that the data is reaching the device, but I suspect the software isn't delineating individual messages.  As previously mentioned, the "received" data continuously increases, but the "successful" numbers do not (screenshots below).

All that said, I'm wondering if this is a futile effort to communicate with the AquaLogic panel.  Even with the AquaLogic display disconnected, it seems the pump, chlorinator, etc. all still function based on the programmed schedule.  I don't see a feasible way that the Nixie would override control of the salt cell.  

I'm starting to think if I wanted to go full Nixie that I'd be better off finding an used AquaRite panel for salt cell control.  That is unless there's a slick way to ditch the salt cell control board all together and control the salt cell via relays from the Nixie using my existing power supply, rectifiers, etc.  I found [this guy's attempt](https://www.troublefreepool.com/threads/how-to-use-my-own-power-supply-for-aquarite-salt-cell.169333/), but it's unclear how this turned out.

Appreciate all the help thus far!

![image](https://github.com/tagyoureit/nodejs-poolController/assets/30021355/8e7ff358-c4cd-442e-a5c9-9c150d43f407)

![image](https://github.com/tagyoureit/nodejs-poolController/assets/30021355/731d8218-593a-4089-9f58-ef1137038b8e)

**Reply by tagyoureit** [reply]
Can you paste some of the output of the `od -x` command ([from the wiki](https://github.com/tagyoureit/nodejs-poolController/wiki/RS-485-Adapter-Details#maclinux)) and paste some of that here.  Make sure no other apps are running or accessing the serial port.  

And if you can also, separately from the above command, attach a [packet capture](https://github.com/tagyoureit/nodejs-poolController/wiki/RS-485-Adapter-Details#maclinux).  I'll take a look.

**Reply by celestinjr** [reply]
Here you go!

[od.log](https://github.com/tagyoureit/nodejs-poolController/files/14938425/od.log)

[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/14938418/replay.zip)

Adding a longer od log:
[od2.log](https://github.com/tagyoureit/nodejs-poolController/files/14938743/od2.log)

**Reply by tagyoureit** [reply]
This data is funky.  It's close, but not exactly, what you need.  Normal Hayward & Aquarite protocols start with [16,2] (0x10,0x02 hex) and end with [16,3] (0x10,0x03 hex).  This as an example packet to a pump:
`0x10, 0x02, 0x0C, 0x01, 0x00, 0x2D, 0x00, 0x4C, 0x10, 0x03 -- Command to pump`

In some places the header and termination bytes appear to be reversed.  EG `0210 0101 1400 0310`.  Other places, the bytes start to look good (the header seems ok) `10 02 01 80 20 00 00 00 00 00 00 b5 00 03 10 02 10 03` but the termination ends with [10,2,10,3] which is wrong.

However, one thing that seems suspicious is that the RS-485 rarely sends text strings.  With Pentair, you can use custom names for the circuits and chlorinator, etc.  But works like "Temp" and "Salt Level" will never appear in the bus.  It almost looks like you are snooping on communications between the board and an LCD that would display this information, and not the RS-485.

**Reply by celestinjr** [reply]
That’s correct. I’m tied into the RS485 remote display port. 

There are three other ports on this board that I see. One is for the wireless antenna for a wireless display/remote. The second is for the local display. I doubt either of these two would show different data than I sent.  The third is labeled J13, but I’m not positive it’s purpose. 

The AquaLogic manual briefly addresses installation of a VS pump or separate AquaRite chlorination panels. All of those connections are supposed to be made into the remote display port that I’m snooping. 

Regarding the ascii messages, I’ve gathered the display text commands have a payload prefix of either [01,03], [01,04], [02,03], or [03,03].

**Reply by tagyoureit** [reply]
It looks like we support AquaLink today, not Aqualogic.  That's something that would need to be added.  

[HomeAssistant has Aqualogic support](https://www.home-assistant.io/integrations/aqualogic/) built in already.   (coming from the [Aqualogic](https://github.com/swilson/aqualogic) github project).
Long thread with more explanation: https://community.home-assistant.io/t/hayward-aqualogic-prologic-automation/52340

We could certainly add this.  It seems to be a limited set of commands/controls.  But I'd suggest starting with one of these and seeing if you can get the panel talking and we can discuss further.

**Reply by celestinjr** [reply]
Roger that.  I appreciate your help.  Currently, I have control implemented in HomeAssistant via the other device that I built, which simply emulates a remote display.  It works fine, but it's a bit laggy and is limited by the AquaLogic functionality of course.  Njspc is appealing in terms of added schedule flexibility, better direct control, and expanded compatibility with non-Hayward equipment options.

I'm working to decide what direction I'm going long term with my equipment, which is aging.  I plan to replace my main pump this summer, so currently working through which automation system I plan to use long term.

I'm happy to contribute back to the project if I go any further down this route.

Thanks again
